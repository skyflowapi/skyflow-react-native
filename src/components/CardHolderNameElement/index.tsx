/*
 Copyright (c) 2022 Skyflow, Inc.
*/
import React, { useEffect, useRef } from "react";
import { Text, TextInput, View } from "react-native";
import type CollectElement from "../../core/CollectElement";
import { CollectElementProps, ElementType, ELEMENT_REQUIRED_ASTERISK, REQUIRED_MARK_DEFAULT_STYLE, ContainerType } from "../../utils/constants";
import SkyflowError from "../../utils/skyflow-error";
import SKYFLOW_ERROR_CODE from "../../utils/skyflow-error-code";
import uuid from 'react-native-uuid';

const CardHolderNameElement: React.FC<CollectElementProps> = ({ container, options = { required: false }, ...rest }) => {
    const [element, setElement] = React.useState<CollectElement>();
    const [elementValue, setElementValue] = React.useState<string>('');
    const [errorText, setErrorText] = React.useState<string>('');
    const [labelStyles, setLabelStyles] = React.useState(rest?.labelStyles?.base || {});
    const [inputStyles, setInputStyles] = React.useState(rest?.inputStyles?.base || {});
    const textInputRef = useRef();
    const uniqueElementID = useRef(uuid.v4() as string);

    useEffect(() => {
        if (container) {
            const element: CollectElement = container.create({ ...rest, type: ElementType.CARDHOLDER_NAME, containerType: container.type }, options);
            setElement(element);
            if (container.type === ContainerType.COLLECT)
                element.setMethods(setErrorText, { setInputStyles: setInputStyles, setLabelStyles: setLabelStyles });
            else if (container.type === ContainerType.COMPOSABLE) {
                element.setMethods(rest.containerMethods.setErrorText, { setInputStyles: setInputStyles, setLabelStyles: setLabelStyles })
                rest.containerMethods.setRef(textInputRef, uniqueElementID.current);
            }
            if (rest.onReady) {
                rest.onReady(element.getClientState());
            }
        } else {
            throw new SkyflowError(SKYFLOW_ERROR_CODE.CONTAINER_OBJECT_IS_REQUIRED, [ElementType.CARDHOLDER_NAME, 'useCollectContainer()'], true)
        }
    }, []);

    return (<View>
        {
            rest.label && (<Text style={labelStyles}>
                {rest.label}
                <Text style={{ ...REQUIRED_MARK_DEFAULT_STYLE, ...rest?.labelStyles?.requiredAsterisk }}>
                    {options.required ? ELEMENT_REQUIRED_ASTERISK : ''}
                </Text>
            </Text>)
        }
        <TextInput
            ref={textInputRef}
            value={elementValue}
            placeholder={rest.placeholder}
            testID={rest?.testID}
            onChangeText={(text) => {
                element?.onChangeElement(text)
                setElementValue(element.getInternalState().value)
            }}
            onFocus={() => {
                element?.onFocusElement()
                setLabelStyles(element.updateLabelStyles());
                setInputStyles(element.updateInputStyles());
            }}
            onBlur={() => {
                element?.onBlurElement();
                if (container.type === ContainerType.COLLECT) {
                    setErrorText(element?.getErrorText() || '');
                } else if (container.type === ContainerType.COMPOSABLE) {
                    rest.containerMethods.setErrorText(element?.getErrorText() || '')
                }
                setElementValue(element.getInternalState().value);
                setLabelStyles(element.updateLabelStyles());
                setInputStyles(element.updateInputStyles());
            }}
            style={inputStyles}
        />

        {
            container && container?.type === ContainerType.COLLECT
            &&
            <Text style={rest?.errorTextStyles?.base || {}} testID="name-error">{errorText}</Text>
        }

    </View>);
}

export default CardHolderNameElement;