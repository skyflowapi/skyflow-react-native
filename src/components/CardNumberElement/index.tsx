/*
 Copyright (c) 2022 Skyflow, Inc.
*/
import React, { useEffect, useRef } from "react";
import { Image, Text, TextInput, View } from "react-native";
import type CollectElement from "../../core/CollectElement";
import { CARD_ENCODED_ICONS, CARD_NUMBER_MASK, CardType, CardTypeValues, DEFAULT_CARD_INPUT_MAX_LENGTH } from "../../core/constants";
import { CollectElementProps, ElementType, ELEMENT_REQUIRED_ASTERISK, REQUIRED_MARK_DEFAULT_STYLE, ContainerType, CARD_NUMBER_ELEMENT_DEFAULT_STYLE, CARD_ICON_DEFAULT_STYLE, IListItem } from "../../utils/constants";
import SkyflowError from "../../utils/skyflow-error";
import SKYFLOW_ERROR_CODE from "../../utils/skyflow-error-code";
import uuid from 'react-native-uuid';
import Dropdown from "../../core/Dropdown";

const CardNumberElement: React.FC<CollectElementProps> = ({ container, options, ...rest }) => {

    const mergedOptions = {
        required: false,
        enableCardIcon: true,
        ...options
    }

    const [element, setElement] = React.useState<CollectElement>(undefined);
    const [elementValue, setElementValue] = React.useState<string>('');
    const [errorText, setErrorText] = React.useState<string>('');
    const [maxLength, setMaxLength] = React.useState<number>(DEFAULT_CARD_INPUT_MAX_LENGTH);
    const [labelStyles, setLabelStyles] = React.useState(rest?.labelStyles?.base || {});
    const [inputStyles, setInputStyles] = React.useState({});
    const textInputRef = useRef();
    const uniqueElementID = useRef(uuid.v4() as string);
    const [cardIcon, setCardIcon] = React.useState<CardType>(CardType.DEFAULT);
    const [cardBrandList, setCardBrandList] = React.useState<IListItem[]>([]);
    const [isDropdownVisible, setIsDropdownVisible] = React.useState<boolean>(false);

    useEffect(() => {
        if (container) {
            const element: CollectElement = container.create({ ...rest, type: ElementType.CARD_NUMBER, containerType: container.type }, mergedOptions);
            setElement(element);
            if (container.type === ContainerType.COLLECT)
                element.setMethods(setErrorText, { setInputStyles: setInputStyles, setLabelStyles: setLabelStyles });
            else if (container.type === ContainerType.COMPOSABLE) {
                element.setMethods(rest.containerMethods.setErrorText, { setInputStyles: setInputStyles, setLabelStyles: setLabelStyles })
                rest.containerMethods.setRef(textInputRef, uniqueElementID.current);
            }
            if (rest.onReady) {
                rest.onReady(element.getClientState());
            }
        } else {
            throw new SkyflowError(SKYFLOW_ERROR_CODE.CONTAINER_OBJECT_IS_REQUIRED, [ElementType.CARD_NUMBER, 'useCollectContainer()'], true)
        }
    }, []);

    React.useEffect(() => {
        if(element && container && (options?.cardMetadata?.scheme?.length>=2)) {
          setIsDropdownVisible(true);
          const cardBrandList: IListItem[] = options.cardMetadata.scheme.map((item: string) => ({
              label: CardTypeValues[item],
              value: item,
            }));
          setCardBrandList(cardBrandList);
          setCardIcon(element?.getCardType())
        } else {
          setIsDropdownVisible(false);
          setCardBrandList([]);
          element?.onDropdownSelect('');
          element?.onChangeElement('', true)
        }
      }, [options?.cardMetadata?.scheme])

    const getCardIconSource = () => {
        if(element && !isDropdownVisible) {
            return CARD_ENCODED_ICONS[element.getCardType()];
        } else if(element && options?.cardMetadata?.scheme) {
            return CARD_ENCODED_ICONS[cardIcon]
        }
    }

    return (<View>
        {
            rest.label && (<Text style={labelStyles}>
                {rest.label}
                <Text style={{ ...REQUIRED_MARK_DEFAULT_STYLE, ...rest?.labelStyles?.requiredAsterisk }}>
                    {mergedOptions.required ? ELEMENT_REQUIRED_ASTERISK : ''}
                </Text>
            </Text>)
        }
        <View style={{...CARD_NUMBER_ELEMENT_DEFAULT_STYLE, flexDirection: 'row', }}>
            {mergedOptions?.enableCardIcon && (
            <Image
              source={getCardIconSource() || CARD_ENCODED_ICONS[CardType.DEFAULT]}
              resizeMode="contain"
              style={{...CARD_ICON_DEFAULT_STYLE, ...rest?.inputStyles?.cardIcon}}
              key={element ? element.getCardType() : CardType.DEFAULT}
            />)}
            
            {mergedOptions?.enableCardIcon && isDropdownVisible && (
                <Dropdown
                 setSelectedValue={(item) => {
                    setCardIcon(item.value)
                    element.onDropdownSelect(item.value)
                    element.onChangeElement('', true)
                 }}
                 listData={cardBrandList}
                 dropdownIconStyles={rest?.inputStyles?.dropdownIcon}
                 dropdownStyles={rest?.inputStyles?.dropdown}
                />
            )}
        <TextInput
            ref={textInputRef}
            value={elementValue}
            placeholder={rest.placeholder}
            onChangeText={(text) => {
                element?.onChangeElement(text)
                setElementValue(element.getInternalState().value)
                setMaxLength(CARD_NUMBER_MASK[element.getCardType()].length || DEFAULT_CARD_INPUT_MAX_LENGTH);
                if (container.type === ContainerType.COMPOSABLE && (!element.getInternalState().isEmpty) && element.getInternalState().isValid) {
                    rest.containerMethods.shiftFocus(uniqueElementID);
                }
            }}
            onFocus={() => {
                element?.onFocusElement();
                setLabelStyles(element.updateLabelStyles());
                setInputStyles(element.updateInputStyles());
            }}
            onBlur={() => {
                element?.onBlurElement();
                if (container.type === ContainerType.COLLECT) {
                    setErrorText(element?.getErrorText() || '');
                } else if (container.type === ContainerType.COMPOSABLE) {
                    rest.containerMethods.setErrorText(element?.getErrorText() || '')
                }
                setElementValue(element.getInternalState().value);
                setLabelStyles(element.updateLabelStyles());
                setInputStyles(element.updateInputStyles());
            }}
            keyboardType='numeric'
            maxLength={maxLength}
            style={{...inputStyles, flex: rest?.inputStyles?.dropdownIcon ? 0 : 1 }}
        />
        </View>
        {
            container && container?.type === ContainerType.COLLECT
            &&
            <Text style={rest?.errorTextStyles?.base || {}}>{errorText}</Text>
        }
    </View>);
}

export default CardNumberElement;