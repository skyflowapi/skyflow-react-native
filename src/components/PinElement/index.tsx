/*
 Copyright (c) 2022 Skyflow, Inc.
*/
import React, { useEffect } from "react";
import { Text, TextInput, View } from "react-native";
import type CollectElement from "../../core/CollectElement";
import { CollectElementProps, ElementType } from "../../utils/constants";
import SkyflowError from "../../utils/skyflow-error";
import SKYFLOW_ERROR_CODE from "../../utils/skyflow-error-code";

const PinElement: React.FC<CollectElementProps> = ({ container, options = { required: false }, ...rest }) => {
    const [element, setElement] = React.useState<CollectElement>();
    const [errorText, setErrorText] = React.useState<string>('');
    const [labelStyles, setLabelStyles] = React.useState(rest?.labelStyles?.base || {});
    const [inputStyles, setInputStyles] = React.useState(rest?.inputStyles?.base || {});

    useEffect(() => {
        if (container) {
            const element: CollectElement = container.create({ ...rest, type: ElementType.PIN }, options);
            setElement(element);
            if (rest.onReady) {
                rest.onReady(element.getInternalState());
            }
        } else {
            throw new SkyflowError(SKYFLOW_ERROR_CODE.CONTAINER_OBJECT_IS_REQUIRED, [ElementType.PIN, 'useCollectContainer()'], true)
        }
    }, []);

    return (<View>
        <Text style={labelStyles}>{rest.label}</Text>
        <TextInput
            placeholder={rest.placeholder}
            onChangeText={(text) => {
                element?.onChangeElement(text)
            }}
            onFocus={() => {
                element?.onFocusElement();
                setLabelStyles(element.updateLabelStyles());
                setInputStyles(element.updateInputStyles());
            }}
            onBlur={() => {
                element?.onBlurElement();
                setErrorText(element?.getErrorText() || '');
                setLabelStyles(element.updateLabelStyles());
                setInputStyles(element.updateInputStyles());
            }}
            maxLength={12}
            keyboardType='numeric'
            style={inputStyles}
        />
        <Text style={rest?.errorTextStyles?.base || {}}>{errorText}</Text>
    </View>);
}

export default PinElement;