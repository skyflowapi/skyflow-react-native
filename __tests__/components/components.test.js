/*
    Copyright (c) 2022 Skyflow, Inc.
*/
import React from 'react';
import InputFieldElement from '../../src/components/InputFieldElement';
import PinElement from '../../src/components/PinElement';
import CvvElement from '../../src/components/CvvElement';
import CardHolderNameElement from '../../src/components/CardHolderNameElement';
import CardNumberElement from '../../src/components/CardNumberElement';
import ExpirationDateElement from '../../src/components/ExpirationDateElement';
import ExpirationYearElement from '../../src/components/ExpirationYearElement';
import ExpirationMonthElement from '../../src/components/ExpirationMonthElement';
import { render, screen, fireEvent } from '@testing-library/react-native';
import CollectContainer from '../../src/core/CollectContainer';
import RevealContainer from '../../src/core/RevealContainer';
import RevealElement from '../../src/components/RevealElement';
import Skyflow from '../../src/core/Skyflow';
import { CardType } from '../../src/core/constants';
import SkyflowProvider from '../../src/components/SkyflowProvider';
import { Text } from 'react-native';
import SkyflowError from '../../src/utils/skyflow-error';
import SKYFLOW_ERROR_CODE from '../../src/utils/skyflow-error-code';
import { ElementType } from '../../src/utils/constants';

const testSkyflowClient = new Skyflow({
  vaultID: '1234',
  vaultURL: 'https://url.com',
  getBearerToken: () => Promise.resolve('valid_token'),
});

const onReadyMock = jest.fn();
const onChangeMock = jest.fn();
const onBlurMock = jest.fn();
const onFocusMock = jest.fn();

const changeTrigger = jest.fn();
const foucsTrigger = jest.fn();
const blurTrigger = jest.fn();

describe('test Collect And Reveal Elements Components', () => {
  let collectContainer;
  beforeEach(() => {
    jest.clearAllMocks();
    jest.resetAllMocks();
    collectContainer = new CollectContainer(testSkyflowClient);
    jest.spyOn(collectContainer, 'create').mockImplementation(() => ({
      getClientState: () => ({}),
      onChangeElement: changeTrigger,
      onFocusElement: foucsTrigger,
      onBlurElement: blurTrigger,
      updateLabelStyles: jest.fn(),
      updateInputStyles: jest.fn(),
      getErrorText: () => '',
      getInternalState: () => ({}),
      getCardType: () => CardType.VISA,
      setMethods: jest.fn(),
    }));
  });

  it('test InputFieldElement component', () => {
    const inputElement = render(
      <InputFieldElement
        table={'table1'}
        column={'string'}
        container={collectContainer}
        onReady={onReadyMock}
        onBlur={onBlurMock}
        onChange={onChangeMock}
        onFocus={onFocusMock}
        label={'First Name'}
        placeholder={'first name'}
        validations={[]}
        options={{
          required: true,
        }}
      />
    );
    expect(inputElement).toMatchSnapshot();
    const textField = screen.getAllByPlaceholderText('first name')[0];
    fireEvent.changeText(textField, 'hello');
    expect(onReadyMock).toBeCalledWith({});
    expect(changeTrigger).toBeCalledWith('hello');

    fireEvent(textField, 'focus');
    expect(foucsTrigger).toBeCalled();

    fireEvent(textField, 'blur');
    expect(blurTrigger).toBeCalled();
    // render without any listeners
    render(
      <InputFieldElement
        table={'table1'}
        column={'string'}
        container={collectContainer}
        label={'First Name'}
        placeholder={'first name'}
        options={{
          required: false,
        }}
      />
    );
    expect(onReadyMock).toBeCalledTimes(1);
    try {
      render(
        <InputFieldElement
          table={'table1'}
          column={'string'}
          placeholder={'first name'}
        />
      );
    } catch (err) {
      expect(err).toEqual(
        new SkyflowError(
          SKYFLOW_ERROR_CODE.CONTAINER_OBJECT_IS_REQUIRED,
          [ElementType.INPUT_FIELD, 'useCollectContainer()'],
          true
        )
      );
    }
  });

  it('test PinElement component', () => {
    const pinElement = render(
      <PinElement
        table={'table1'}
        column={'string'}
        container={collectContainer}
        onReady={onReadyMock}
        onBlur={onBlurMock}
        onChange={onChangeMock}
        onFocus={onFocusMock}
        label={'Pin'}
        placeholder={'pin'}
        validations={[]}
        options={{
          required: true,
        }}
      />
    );
    expect(pinElement).toMatchSnapshot();
    const textField = screen.getAllByPlaceholderText('pin')[0];
    fireEvent.changeText(textField, '1234');
    expect(onReadyMock).toBeCalledWith({});
    expect(changeTrigger).toBeCalledWith('1234');

    fireEvent(textField, 'focus');
    expect(foucsTrigger).toBeCalled();

    fireEvent(textField, 'blur');
    expect(blurTrigger).toBeCalled();
    // render without any listeners
    render(
      <PinElement
        table={'table1'}
        column={'string'}
        container={collectContainer}
        label={'Pin'}
        placeholder={'pin'}
        options={{
          required: false,
        }}
      />
    );
    expect(onReadyMock).toHaveBeenCalledTimes(1);
    try {
      render(
        <PinElement table={'table1'} column={'string'} placeholder={'pin'} />
      );
    } catch (err) {
      expect(err).toEqual(
        new SkyflowError(
          SKYFLOW_ERROR_CODE.CONTAINER_OBJECT_IS_REQUIRED,
          [ElementType.PIN, 'useCollectContainer()'],
          true
        )
      );
    }
  });

  it('test CvvElement component', () => {
    const cvvElement = render(
      <CvvElement
        table={'table1'}
        column={'string'}
        container={collectContainer}
        onReady={onReadyMock}
        onBlur={onBlurMock}
        onChange={onChangeMock}
        onFocus={onFocusMock}
        label={'Cvv'}
        placeholder={'cvv'}
        validations={[]}
        options={{
          required: true,
        }}
      />
    );
    expect(cvvElement).toMatchSnapshot();
    const textField = screen.getAllByPlaceholderText('cvv')[0];
    fireEvent.changeText(textField, '123');
    expect(onReadyMock).toBeCalledWith({});
    expect(changeTrigger).toBeCalledWith('123');

    fireEvent(textField, 'focus');
    expect(foucsTrigger).toBeCalled();

    fireEvent(textField, 'blur');
    expect(blurTrigger).toBeCalled();
    // render without any listeners
    render(
      <CvvElement
        table={'table1'}
        column={'string'}
        container={collectContainer}
        label={'Cvv'}
        placeholder={'cvv'}
        options={{
          required: false,
        }}
      />
    );
    expect(onReadyMock).toHaveBeenCalledTimes(1);
    try {
      render(
        <CvvElement table={'table1'} column={'string'} placeholder={'cvv'} />
      );
    } catch (err) {
      expect(err).toEqual(
        new SkyflowError(
          SKYFLOW_ERROR_CODE.CONTAINER_OBJECT_IS_REQUIRED,
          [ElementType.CVV, 'useCollectContainer()'],
          true
        )
      );
    }
  });

  it('test CardHolderNameElement component', () => {
    const cardHolderElement = render(
      <CardHolderNameElement
        table={'table1'}
        column={'string'}
        container={collectContainer}
        onReady={onReadyMock}
        onBlur={onBlurMock}
        onChange={onChangeMock}
        onFocus={onFocusMock}
        label={'CardHolderName'}
        placeholder={'CardHolder Name'}
        validations={[]}
        options={{
          required: true,
        }}
      />
    );
    expect(cardHolderElement).toMatchSnapshot();
    const textField = screen.getAllByPlaceholderText('CardHolder Name')[0];
    fireEvent.changeText(textField, 'test');
    expect(onReadyMock).toBeCalledWith({});
    expect(changeTrigger).toBeCalledWith('test');

    fireEvent(textField, 'focus');
    expect(foucsTrigger).toBeCalled();

    fireEvent(textField, 'blur');
    expect(blurTrigger).toBeCalled();
    // render without any listeners
    render(
      <CardHolderNameElement
        table={'table1'}
        column={'string'}
        label={'CardHolder Name'}
        container={collectContainer}
        placeholder={'CardHolder Name'}
        options={{
          required: false,
        }}
      />
    );
    expect(onReadyMock).toHaveBeenCalledTimes(1);
    try {
      render(
        <CardHolderNameElement
          table={'table1'}
          column={'string'}
          placeholder={'CardHolder Name'}
        />
      );
    } catch (err) {
      expect(err).toEqual(
        new SkyflowError(
          SKYFLOW_ERROR_CODE.CONTAINER_OBJECT_IS_REQUIRED,
          [ElementType.CARDHOLDER_NAME, 'useCollectContainer()'],
          true
        )
      );
    }
  });

  it('test CardNumberElement component', () => {
    const cardNumberElement = render(
      <CardNumberElement
        table={'table1'}
        column={'string'}
        container={collectContainer}
        onReady={onReadyMock}
        onBlur={onBlurMock}
        onChange={onChangeMock}
        onFocus={onFocusMock}
        label={'CardNumber'}
        placeholder={'card number'}
        validations={[]}
        options={{
          required: true,
        }}
      />
    );
    expect(cardNumberElement).toMatchSnapshot();
    const textField = screen.getAllByPlaceholderText('card number')[0];
    fireEvent.changeText(textField, '4111111111111111');
    expect(onReadyMock).toBeCalledWith({});
    expect(changeTrigger).toBeCalledWith('4111111111111111');

    fireEvent(textField, 'focus');
    expect(foucsTrigger).toBeCalled();

    fireEvent(textField, 'blur');
    expect(blurTrigger).toBeCalled();

    const collectContainerTest = new CollectContainer(testSkyflowClient);
    jest.spyOn(collectContainerTest, 'create').mockImplementation(() => ({
      getInternalState: () => ({}),
      getCardType: () => null,
      setMethods: jest.fn(),
    }));
    // render without any listeners
    render(
      <CardNumberElement
        table={'table1'}
        column={'string'}
        container={collectContainerTest}
        label={'Card Number'}
        placeholder={'card number'}
        options={{
          required: false,
        }}
      />
    );
    expect(onReadyMock).toHaveBeenCalledTimes(1);

    try {
      render(
        <CardNumberElement
          table={'table1'}
          column={'string'}
          placeholder={'card number'}
        />
      );
    } catch (err) {
      expect(err).toEqual(
        new SkyflowError(
          SKYFLOW_ERROR_CODE.CONTAINER_OBJECT_IS_REQUIRED,
          [ElementType.CARD_NUMBER, 'useCollectContainer()'],
          true
        )
      );
    }
  });

  it('test ExpirationDateElement component', () => {
    const expirationDateElement = render(
      <ExpirationDateElement
        table={'table1'}
        column={'string'}
        container={collectContainer}
        onReady={onReadyMock}
        onBlur={onBlurMock}
        onChange={onChangeMock}
        onFocus={onFocusMock}
        label={'Expiration Date'}
        placeholder={'expiration date'}
        validations={[]}
        options={{
          required: true,
        }}
      />
    );
    expect(expirationDateElement).toMatchSnapshot();
    const textField = screen.getAllByPlaceholderText('expiration date')[0];
    fireEvent.changeText(textField, '12/2032');
    expect(onReadyMock).toBeCalledWith({});
    expect(changeTrigger).toBeCalledWith('12/2032');

    fireEvent(textField, 'focus');
    expect(foucsTrigger).toBeCalled();

    fireEvent(textField, 'blur');
    expect(blurTrigger).toBeCalled();

    // render without any listeners
    render(
      <ExpirationDateElement
        table={'table1'}
        column={'string'}
        container={collectContainer}
        label={'Expiration Date'}
        placeholder={'expiration date'}
        options={{
          required: false,
        }}
      />
    );
    expect(onReadyMock).toHaveBeenCalledTimes(1);

    try {
      render(
        <ExpirationDateElement
          table={'table1'}
          column={'string'}
          placeholder={'expiration date'}
        />
      );
    } catch (err) {
      expect(err).toEqual(
        new SkyflowError(
          SKYFLOW_ERROR_CODE.CONTAINER_OBJECT_IS_REQUIRED,
          [ElementType.EXPIRATION_DATE, 'useCollectContainer()'],
          true
        )
      );
    }
  });

  it('test ExpirationMonthElement component', () => {
    const expirationMonthElement = render(
      <ExpirationMonthElement
        table={'table1'}
        column={'string'}
        container={collectContainer}
        onReady={onReadyMock}
        onBlur={onBlurMock}
        onChange={onChangeMock}
        onFocus={onFocusMock}
        label={'Expiration Month'}
        placeholder={'expiration month'}
        validations={[]}
        options={{
          required: true,
        }}
      />
    );
    expect(expirationMonthElement).toMatchSnapshot();
    const textField = screen.getAllByPlaceholderText('expiration month')[0];
    fireEvent.changeText(textField, '12');
    expect(onReadyMock).toBeCalledWith({});
    expect(changeTrigger).toBeCalledWith('12');

    fireEvent(textField, 'focus');
    expect(foucsTrigger).toBeCalled();

    fireEvent(textField, 'blur');
    expect(blurTrigger).toBeCalled();

    // render without any listeners
    render(
      <ExpirationMonthElement
        table={'table1'}
        column={'string'}
        container={collectContainer}
        label={'Expiration Month'}
        placeholder={'expiration month'}
        options={{
          required: false,
        }}
      />
    );
    expect(onReadyMock).toHaveBeenCalledTimes(1);
    try {
      render(
        <ExpirationMonthElement
          table={'table1'}
          column={'string'}
          placeholder={'expiration month'}
        />
      );
    } catch (err) {
      expect(err).toEqual(
        new SkyflowError(
          SKYFLOW_ERROR_CODE.CONTAINER_OBJECT_IS_REQUIRED,
          [ElementType.EXPIRATION_MONTH, 'useCollectContainer()'],
          true
        )
      );
    }
  });

  it('test ExpirationYearElement component', () => {
    const expirationYearElement = render(
      <ExpirationYearElement
        table={'table1'}
        column={'string'}
        container={collectContainer}
        onReady={onReadyMock}
        onBlur={onBlurMock}
        onChange={onChangeMock}
        onFocus={onFocusMock}
        label={'Expiration Year'}
        placeholder={'expiration year'}
        validations={[]}
        options={{
          required: true,
        }}
      />
    );
    expect(expirationYearElement).toMatchSnapshot();
    const textField = screen.getAllByPlaceholderText('expiration year')[0];
    fireEvent.changeText(textField, '2032');
    expect(onReadyMock).toBeCalledWith({});
    expect(changeTrigger).toBeCalledWith('2032');

    fireEvent(textField, 'focus');
    expect(foucsTrigger).toBeCalled();

    fireEvent(textField, 'blur');
    expect(blurTrigger).toBeCalled();

    // render without any listeners
    render(
      <ExpirationYearElement
        table={'table1'}
        column={'string'}
        container={collectContainer}
        label={'Expiration Year'}
        placeholder={'expiration year'}
        options={{
          required: false,
        }}
      />
    );
    expect(onReadyMock).toHaveBeenCalledTimes(1);
    try {
      render(
        <ExpirationYearElement
          table={'table1'}
          column={'string'}
          placeholder={'expiration year'}
        />
      );
    } catch (err) {
      expect(err).toEqual(
        new SkyflowError(
          SKYFLOW_ERROR_CODE.CONTAINER_OBJECT_IS_REQUIRED,
          [ElementType.EXPIRATION_YEAR, 'useCollectContainer()'],
          true
        )
      );
    }
  });

  it('test RevealElement component', () => {
    const revealSetMethodMock = jest.fn();
    const revealContainer = new RevealContainer(testSkyflowClient);
    jest.spyOn(revealContainer, 'create').mockImplementation(() => ({
      setMethods: revealSetMethodMock,
    }));

    const revealElement = render(
      <RevealElement
        token={'test_token'}
        container={revealContainer}
        label={'Card Number'}
        altText={'XXXX XXXX XXXX XXXX'}
      />
    );

    expect(revealElement).toMatchSnapshot();
    expect(revealSetMethodMock).toBeCalledTimes(1);

    // render without alttext
    const revealElement2 = render(
      <RevealElement
        token={'test_token'}
        container={revealContainer}
        label={'Card Number'}
      />
    );
    try {
      render(<RevealElement token={'test_token'} label={'Card Number'} />);
    } catch (err) {
      expect(err).toEqual(
        new SkyflowError(
          SKYFLOW_ERROR_CODE.CONTAINER_OBJECT_IS_REQUIRED,
          ['Reveal', 'useRevealContainer()'],
          true
        )
      );
    }
  });

  it('test skyflow provider', () => {
    const testSkyflowConfig = {
      vaultID: '1234',
      vaultURL: 'https://vaulturl.com',
      getBearerToken: () => Promise.resolve('valid_token'),
    };

    const providerElement = render(
      <SkyflowProvider config={testSkyflowConfig}>
        <Text>Provider Childern</Text>
      </SkyflowProvider>
    );
    expect(providerElement).toMatchSnapshot();
  });
});
