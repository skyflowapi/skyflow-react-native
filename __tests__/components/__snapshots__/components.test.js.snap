// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Collect And Reveal Elements Components test CardHolderNameElement component 1`] = `
<View>
  <Text
    style={Object {}}
  >
    CardHolderName
    <Text
      style={
        Object {
          "color": "red",
        }
      }
    >
       *
    </Text>
  </Text>
  <TextInput
    onBlur={[Function]}
    onChangeText={[Function]}
    onFocus={[Function]}
    placeholder="CardHolder Name"
    style={Object {}}
    value=""
  />
  <Text
    style={Object {}}
  />
</View>
`;

exports[`test Collect And Reveal Elements Components test CardNumberElement component 1`] = `
<View>
  <Text
    style={Object {}}
  >
    CardNumber
    <Text
      style={
        Object {
          "color": "red",
        }
      }
    >
       *
    </Text>
  </Text>
  <View
    style={
      Object {
        "borderColor": "#eae8ee",
        "borderRadius": 4,
        "borderWidth": 2,
        "flexDirection": "row",
        "gap": 4,
        "paddingHorizontal": 6,
      }
    }
  >
    <Image
      resizeMode="contain"
      source={
        Object {
          "testUri": "../../../assets/visa.png",
        }
      }
      style={
        Object {
          "height": 50,
          "width": 50,
        }
      }
    />
    <TextInput
      keyboardType="numeric"
      maxLength={23}
      onBlur={[Function]}
      onChangeText={[Function]}
      onFocus={[Function]}
      placeholder="card number"
      style={
        Object {
          "flex": 1,
        }
      }
      value=""
    />
  </View>
  <Text
    style={Object {}}
  />
</View>
`;

exports[`test Collect And Reveal Elements Components test CvvElement component 1`] = `
<View>
  <Text
    style={Object {}}
  >
    Cvv
    <Text
      style={
        Object {
          "color": "red",
        }
      }
    >
       *
    </Text>
  </Text>
  <TextInput
    keyboardType="numeric"
    maxLength={4}
    onBlur={[Function]}
    onChangeText={[Function]}
    onFocus={[Function]}
    placeholder="cvv"
    style={Object {}}
    value=""
  />
  <Text
    style={Object {}}
  />
</View>
`;

exports[`test Collect And Reveal Elements Components test ExpirationDateElement component 1`] = `
<View>
  <Text
    style={Object {}}
  >
    Expiration Date
    <Text
      style={
        Object {
          "color": "red",
        }
      }
    >
       *
    </Text>
  </Text>
  <TextInput
    keyboardType="numeric"
    maxLength={5}
    onBlur={[Function]}
    onChangeText={[Function]}
    onFocus={[Function]}
    placeholder="expiration date"
    style={Object {}}
    value=""
  />
  <Text
    style={Object {}}
  />
</View>
`;

exports[`test Collect And Reveal Elements Components test ExpirationMonthElement component 1`] = `
<View>
  <Text
    style={Object {}}
  >
    Expiration Month
    <Text
      style={
        Object {
          "color": "red",
        }
      }
    >
       *
    </Text>
  </Text>
  <TextInput
    keyboardType="numeric"
    maxLength={2}
    onBlur={[Function]}
    onChangeText={[Function]}
    onFocus={[Function]}
    placeholder="expiration month"
    style={Object {}}
    value=""
  />
  <Text
    style={Object {}}
  />
</View>
`;

exports[`test Collect And Reveal Elements Components test ExpirationYearElement component 1`] = `
<View>
  <Text
    style={Object {}}
  >
    Expiration Year
    <Text
      style={
        Object {
          "color": "red",
        }
      }
    >
       *
    </Text>
  </Text>
  <TextInput
    keyboardType="numeric"
    maxLength={2}
    onBlur={[Function]}
    onChangeText={[Function]}
    onFocus={[Function]}
    placeholder="expiration year"
    style={Object {}}
    value=""
  />
  <Text
    style={Object {}}
  />
</View>
`;

exports[`test Collect And Reveal Elements Components test InputFieldElement component 1`] = `
<View>
  <Text
    style={Object {}}
  >
    First Name
    <Text
      style={
        Object {
          "color": "red",
        }
      }
    >
       *
    </Text>
  </Text>
  <TextInput
    onBlur={[Function]}
    onChangeText={[Function]}
    onFocus={[Function]}
    placeholder="first name"
    style={Object {}}
    value=""
  />
  <Text
    style={Object {}}
  />
</View>
`;

exports[`test Collect And Reveal Elements Components test PinElement component 1`] = `
<View>
  <Text
    style={Object {}}
  >
    Pin
    <Text
      style={
        Object {
          "color": "red",
        }
      }
    >
       *
    </Text>
  </Text>
  <TextInput
    keyboardType="numeric"
    maxLength={12}
    onBlur={[Function]}
    onChangeText={[Function]}
    onFocus={[Function]}
    placeholder="pin"
    style={Object {}}
    value=""
  />
  <Text
    style={Object {}}
  />
</View>
`;

exports[`test Collect And Reveal Elements Components test RevealElement component 1`] = `
Array [
  <Text
    style={Object {}}
  >
    Card Number
  </Text>,
  <Text
    selectable={true}
    style={Object {}}
  >
    XXXX XXXX XXXX XXXX
  </Text>,
  <Text
    style={Object {}}
  />,
]
`;

exports[`test Collect And Reveal Elements Components test skyflow provider 1`] = `
<Text>
  Provider Childern
</Text>
`;
